wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java

```diff
@@ -0,0 +1,1587 @@
+                        checkoutCommand.setStartPoint(remoteBranch.get().getName());
+                    }
+                }
+            }
+            if (trackBranch != null) {
+                if (name == null) {
+                    checkoutCommand.setName(cleanRemoteName(trackBranch));
+                }
+                checkoutCommand.setCreateBranch(true);
+                checkoutCommand.setStartPoint(trackBranch);
+            }
+            checkoutCommand.setUpstreamMode(SetupUpstreamMode.SET_UPSTREAM);
+        }
+        try {
+            checkoutCommand.call();
+        } catch (GitAPIException exception) {
+            if (exception.getMessage().endsWith("already exists")) {
+                throw new GitException(String.format(ERROR_BRANCH_NAME_EXISTS, name != null ? name : cleanRemoteName(trackBranch)));
+            }
+            throw new GitException(exception.getMessage(), exception);
+        }
+    }
+
+    @Override
+    public Branch branchCreate(BranchCreateRequest request) throws GitException {
+        CreateBranchCommand createBranchCommand = getGit().branchCreate().setName(request.getName());
+        String start = request.getStartPoint();
+        if (start != null) {
+            createBranchCommand.setStartPoint(start);
+        }
+        try {
+            Ref brRef = createBranchCommand.call();
+            String refName = brRef.getName();
+            String displayName = Repository.shortenRefName(refName);
+            return newDto(Branch.class).withName(refName)
+                                       .withDisplayName(displayName)
+                                       .withActive(false)
+                                       .withRemote(false);
+        } catch (GitAPIException exception) {
+            throw new GitException(exception.getMessage(), exception);
+        }
+    }
+
+    @Override
+    public void branchDelete(BranchDeleteRequest request) throws GitException {
+        try {
+            getGit().branchDelete()
+                    .setBranchNames(request.getName())
+                    .setForce(request.isForce())
+                    .call();
+        } catch (GitAPIException exception) {
+            throw new GitException(exception.getMessage(), exception);
+        }
+    }
+
+    @Override
+    public void branchRename(String oldName, String newName) throws GitException {
+        try {
+            getGit().branchRename()
+                    .setOldName(oldName)
+                    .setNewName(newName)
+                    .call();
+        } catch (GitAPIException exception) {
+            throw new GitException(exception.getMessage(), exception);
+        }
+    }
+
+    @Override
+    public List<Branch> branchList(BranchListRequest request) throws GitException {
+        String listMode = request.getListMode();
+        if (!(listMode == null || listMode.equals(BranchListRequest.LIST_ALL) || listMode.equals(BranchListRequest.LIST_REMOTE))) {
```

Comment:
We can not avoid `listMode == null` here because we have ability to use `BranchListRequest` without specifying listMode, so we have to check `listMode' for `null`. Flipped equals conditions to make it more readable.